In [2]:
import numpy as np

## Eigenvalue and Eigenvectors

* For a n by n matrix, if there exists a scala $\lambda$ and a non-zero vector $\vec{v}$ such at $A\vec{x}$ = $\lambda\vec{x}$, then $\lambda$ is $A$'s eigenvalue and $\vec{v}$ is the corresponding eigenvectors
* If $\lambda$ is $A$'s eigenvalue, then there are infinitely many eigenvectors

## Calculating Eigenvalue

* For a 2 by 2 matrix, det($A$ - $\lambda{I_2}$) = 0
* det($A$ - $\lambda{I_2}$) = $\lambda^2$ - $tr(A)\lambda$ + $det(A)$ = 0

## Similar Matrices

* For an invertible matrix $S$, if $S^{-1} A S$ = $B$, then square matrices A and B are called **similar**
* Similar matrices share same eigenvalues

## Matrix Diagonalization

* If a square matrix $A$ is similar to a diagonal matrix $D$, which means $S^{-1} A S$ = $D$, then $A$ is said to be **diagonalizable**
* This means that a n by n matrix $A$ has to have n linearly independent eigenvectors in order to be diagonalizable
* Diagonalizing a n by n matrix $A$ with n linearly independent eigenvectors:
    * Find $A$'s n linearly independent eigenvectors $\vec{s_1}$, $\vec{s_2}$, ..., $\vec{s_n}$
    * Form matrix $S$ = [$\vec{s_1}$,, $\vec{s_2}$, ..., $\vec{s_n}$]
    * The resulting matrix of $S^{-1} A S$ will be a diagonal matrix whose diagonal elements will be the corresponding eigenvalues

## Practice with Codes

Use Python's NumPy to find the followings:<br>
1. Handle numpy.linalg library 
2. Find eigenvectors

### 1. Finding Eigenvector

This is also refered to as **eigen decomposition**, which is the process of finding eigenvalues and eigenvectors

In [3]:
#create a 3x3 array
A = np.array([[1,2,3], [4,5,6], [7,8,9]])
A

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [4]:
np.linalg.eig(A)

(array([ 1.61168440e+01, -1.11684397e+00, -1.30367773e-15]),
 array([[-0.23197069, -0.78583024,  0.40824829],
        [-0.52532209, -0.08675134, -0.81649658],
        [-0.8186735 ,  0.61232756,  0.40824829]]))

* In ndarray, the columns are eigenvectors, not the rows. For example column (-0.23197069, -0.52532209, -0.8186735) is one eigenvector

In [17]:
eigen_vec1 = eigen_vec[:, 0]
print("1st eigenvector : " + str(eigen_vec1))

eigen_value1=eigen_val[0]
eigen_value1
print("1st eigenvalue : " + str(eigen_value1))

print("Dot product of matrix A and 1st eigenvector : " + str(A.dot(eigen_vec1)))
print("Product of 1st eigenvalue and 1st eigenvector : " + str(eigen_value1 * eigen_vec1))

1st eigenvector : [-0.23197069 -0.52532209 -0.8186735 ]
1st eigenvalue : 16.116843969807043
Dot product of matrix A and 1st eigenvector : [ -3.73863537  -8.46653421 -13.19443305]
Product of 1st eigenvalue and 1st eigenvector : [ -3.73863537  -8.46653421 -13.19443305]


* Here we confirmed that there exists a eigenvector such that $A\vec{x} = \lambda\vec{x}$

### 2. Eigenvectors Are Unit Vectors (Size is 1)

In [5]:
eigen_val, eigen_vec = np.linalg.eig(A)
eigen_vec ** 2

array([[0.0538104 , 0.61752916, 0.16666667],
       [0.2759633 , 0.00752579, 0.66666667],
       [0.6702263 , 0.37494504, 0.16666667]])

In [6]:
#each eigenvectors' length is 1
np.sum(eigen_vec ** 2, axis=0)

array([1., 1., 1.])

### 3. Eigenvectors Are Orthogonal ONLY for Symmetric Matrix

In [27]:
# A는 symmetric이 아니다.
print("Matrix A is NOT symmetric")
display(A)
print("========================")
print("Therefore, A's eigenvectors are NOT orthogonal : " + str(np.dot(eigen_vec[:,0], eigen_vec[:,1])))

Matrix A is NOT symmetric


array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

Therefore, A's eigenvectors are NOT orthogonal : -0.27343437080986494


In [29]:
# B는 symmetric 이다.
B = np.array([[1, -2], [-2, 1]])
print("Matrix B IS symmetric")
display(B)
print("========================")
b1, b2 = np.linalg.eig(B)

print("Therefore, B's eigenvectors ARE orthogonal : " + str(np.dot(b2[:,0], b2[:,1])))

Matrix B IS symmetric


array([[ 1, -2],
       [-2,  1]])

Therefore, B's eigenvectors ARE orthogonal : 0.0


### 4. Diagonalization

* If a square matrix $A$ and a diagonal matrix $D$ are similar matrices such that $S^{-1}AS = D$, then $A$ is said to be **diagonalizable**.
* Requirement is that a n by n matrix must have n linearly independent eigenvectors
* How to diagonalize a n by n matrix with n linearly independent eigenvectors:<br>
    1. Find $A$'s $n$ linearly independent eigenvectors $\vec{s_1}, \vec{s_2}, ..., \vec{s_n}$
    2. Matrix $S$ = [$\vec{s_1} \dots \vec{s_n}$]
    3. The product matrix of $S^{-1}AS$ will be a diagonal matrix whose diagonal elements are corresponding eigenvalues of $\vec{s_1} \dots \vec{s_n}$


* Because this is so important, **remember**
    1. Diagonalization only exists for **square** matrices
    2. Eigenvectors must be linearly independent in order for the matrix to be diagonalized
    3. **Symmetric matrices** are diagonalizable

In [45]:
#create a symmetric matrix
c = np.array([[1, 2, 3], [2, 4, 5], [3, 5, 6]])

print("Symmetric matrix C")
display(c)

Symmetric matrix C


array([[1, 2, 3],
       [2, 4, 5],
       [3, 5, 6]])

In [46]:
#find eigenvalue and eigenvectors of matrix c
eigen_val, eigen_vec = np.linalg.eig(c)

print("Eigen Values")
print(eigen_val)
print("========================================")
print("Eigen Vectors")
print(eigen_vec)

Eigen Values
[11.34481428 -0.51572947  0.17091519]
Eigen Vectors
[[-0.32798528 -0.73697623  0.59100905]
 [-0.59100905 -0.32798528 -0.73697623]
 [-0.73697623  0.59100905  0.32798528]]


* For symmetric matrices, $S^{-1}AS = D$, where the diagonal elements of $D$ are the eigenvalues of $A$<br>
* Put it differently, $A = SDS^{-1}$<br>
* Let's check if this is true

In [54]:
eigen_vec.dot(np.diag(eigen_val).dot(np.linalg.inv(eigen_vec)))

array([[1., 2., 3.],
       [2., 4., 5.],
       [3., 5., 6.]])